File: sdk/rust/src/wire.rs
use serde::{Deserialize, Serialize};

/// One entry of welcome's `peers` array.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WirePeer {
    pub id: String,
    pub static_key: String,
}

/// The JSON frame exchanged on the control channel (text frames only). Fields are
/// selectively populated depending on `msg_type`; unknown fields on the way in are
/// ignored by serde, and `None` fields are omitted going out.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct WireMessage {
    #[serde(rename = "type")]
    pub msg_type: String,

    // welcome
    #[serde(skip_serializing_if = "Option::is_none")]
    pub protocol_version: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub device_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub peers: Option<Vec<WirePeer>>,

    // announce_key
    #[serde(skip_serializing_if = "Option::is_none")]
    pub static_key: Option<String>,

    // pair_code / pair_accept / pair_complete (code doubles as the error machine code
    // on error messages, matching the server's wire shape)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expires_in: Option<u64>,

    // pair_complete (peer_id is reused by peer_status)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub peer_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub peer_static_key: Option<String>,

    // peer_status
    #[serde(skip_serializing_if = "Option::is_none")]
    pub online: Option<bool>,

    // error
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
}

impl WireMessage {
    pub fn announce_key(static_key: &str) -> Self {
        Self { msg_type: "announce_key".into(), static_key: Some(static_key.into()), ..Default::default() }
    }
    pub fn pair_request() -> Self {
        Self { msg_type: "pair_request".into(), ..Default::default() }
    }
    pub fn pair_accept(code: &str) -> Self {
        Self { msg_type: "pair_accept".into(), code: Some(code.into()), ..Default::default() }
    }
    pub fn ping() -> Self {
        Self { msg_type: "ping".into(), ..Default::default() }
    }
}

File: sdk/go/wire.go
package relayly

// Wire control-message type constants (docs/PROTOCOL.md §5). The old "auth"/"send"
// JSON types and their payload/nonce fields are gone: auth is now HTTP-layer query
// params, and encrypted data travels as a binary E2E envelope, never JSON.
const (
	msgTypeWelcome      = "welcome"
	msgTypeAnnounceKey  = "announce_key"
	msgTypePairRequest  = "pair_request"
	msgTypePairCode     = "pair_code"
	msgTypePairAccept   = "pair_accept"
	msgTypePairComplete = "pair_complete"
	msgTypePeerStatus   = "peer_status"
	msgTypePing         = "ping"
	msgTypePong         = "pong"
	msgTypeError        = "error"
)

// protocolVersion is the wire protocol version this SDK implements (docs/PROTOCOL.md).
const protocolVersion = 1

// wirePeer is one entry of welcome's peers array.
type wirePeer struct {
	ID        string `json:"id"`
	StaticKey string `json:"static_key"`
}

// controlMessage is the JSON frame exchanged on the control channel (text frames
// only). Fields are selectively populated depending on Type; unknown fields on the
// way in are ignored by json.Unmarshal, and zero-value fields are omitted going out.
type controlMessage struct {
	Type string `json:"type"`

	// welcome
	ProtocolVersion int        `json:"protocol_version,omitempty"`
	DeviceID        string     `json:"device_id,omitempty"`
	Peers           []wirePeer `json:"peers,omitempty"`

	// announce_key
	StaticKey string `json:"static_key,omitempty"`

	// pair_code / pair_accept / pair_complete (Code doubles as the error machine
	// code on error messages, matching the server's wire shape)
	Code      string `json:"code,omitempty"`
	ExpiresIn int    `json:"expires_in,omitempty"`

	// pair_complete (PeerID is reused by peer_status)
	PeerID        string `json:"peer_id,omitempty"`
	PeerStaticKey string `json:"peer_static_key,omitempty"`

	// peer_status — a pointer so an explicit "online":false is distinguishable from
	// "field absent" while decoding.
	Online *bool `json:"online,omitempty"`

	// error
	Message string `json:"message,omitempty"`
}
